cirq-core/cirq/ops/gate_operation_test.py

```diff
@@ -465,3 +468,78 @@ def qubit_index_to_equivalence_group_key(self, index: int) -> int:
     assert gate(qubits[0], qubits[1], qubits[2], qubits[3]) == gate(
         qubits[3], qubits[1], qubits[2], qubits[0]
     )
+
+
+def test_gate_to_operation_to_gate_round_trips():
+    def all_subclasses(cls):
+        return set(cls.__subclasses__()).union(
+            [s for c in cls.__subclasses__() for s in all_subclasses(c)]
+        )
```

Comment:
This isn't at risk for infinite recursion right ? (if we had two gates with circular dependencies of some kind)

Reply:
I think this is fine since the type hierarchy has to be a dag.